runtime/rasdump/dmpsup.c

```diff
@@ -532,10 +539,32 @@ configureDumpAgents(J9JavaVM *vm)
 	/* Process IBM_XE_COE_NAME */
 	mapDumpSettings(vm, agentOpts, &agentNum);
 	
-	/* Process -Xdump command-line options (L..R)*/
+
+	/*
+	 * Process -XX:[+-]HeapDumpOnOutOfMemoryError.
+	 */
+	disableIndex = FIND_AND_CONSUME_ARG(EXACT_MATCH, VMOPT_XXNOHEAPDUMPONOOM, NULL);
+	enableIndex = FIND_AND_CONSUME_ARG(EXACT_MATCH, VMOPT_XXHEAPDUMPONOOM, NULL);
+	processXXHeapDump = ((enableIndex >= 0) || (disableIndex >= 0));
+
+	/*
+	 * Process -Xdump command-line options (L..R).
+	 * Treat -XX:[+-]HeapDumpOnOutOfMemoryError as an alias of -Xdump.
+	 */
 	xdumpIndex = FIND_ARG_IN_VMARGS_FORWARD(OPTIONAL_LIST_MATCH, VMOPT_XDUMP, NULL);
 	while ( xdumpIndex >= 0 )
 	{
+		if (processXXHeapDump) {
+			if ((enableIndex > disableIndex) && (enableIndex < xdumpIndex)) {
+				enableDumpOnOutOfMemoryError(vm, agentOpts, &agentNum);
+				processXXHeapDump = FALSE;
+				continue;
+			} else if ((disableIndex > enableIndex) && (disableIndex < xdumpIndex)) {
+				disableDumpOnOutOfMemoryError(vm, agentOpts, agentNum);
+				processXXHeapDump = FALSE;
+				continue;
+			}
+		}
```

Comment:
The interaction between `-XX` options and `-Xdump` options doesn't seem right to me: it seems to ignore the first `-Xdump` option if either `-XX:[+-]HeapDumpOnOutOfMemoryError` appears before the current option.

Reply:
That's a fair question.

Assume we have the following: 
`-Xdump:foo -XX:[+]HeapDumpOnOutOfMemoryError -Xdump:bar`, which have argument indices 1, 2, 3 respectively. 

On the first iteration we process `-Xdump:foo` because `xdumpIndex` (i.e. 1) <`enableIndex` (i.e. 2).

Next we process `-Xdump:bar`, `xdumpIndex`=3. We process the `HeapDumpOnOutOfMemoryError` argument because `processXXHeapDump` is true and `enableIndex ` (i.e. 2) <`xdumpIndex`. We clear `HeapDumpOnOutOfMemoryError` and `continue`, skipping the `FIND_NEXT_ARG_IN_VMARGS_FORWARD`. 

The next time through the loop, `xdumpIndex`=3 (still) but `processXXHeapDump` is false, so we process `-Xdump:bar`.

By way of experiment:
`-Xdump:none  -Xdump:system:events=systhrow,filter=java/lang/AbstractMethodError -XX:+HeapDumpOnOutOfMemoryError -Xdump:java:events=systhrow,filter=java/lang/AbstractMethodError -Xdump:what HelloWorld`
gives (trimmed for brevity):
```
Registered dump agents
----------------------
-Xdump:system:  /* first -Xdump argument */
    events=systhrow,
    filter=java/lang/AbstractMethodError,
----------------------
-Xdump:system:
    events=systhrow,
    filter=java/lang/OutOfMemoryError,
----------------------
-Xdump:heap:
    events=systhrow,
    filter=java/lang/OutOfMemoryError,
    opts=PHD
----------------------
-Xdump:java:
    events=systhrow,
    filter=java/lang/OutOfMemoryError,
----------------------
-Xdump:java:  /* last -Xdump argument */
    events=systhrow,
    filter=java/lang/AbstractMethodError,
----------------------
-Xdump:snap:
    events=systhrow,
    filter=java/lang/OutOfMemoryError,
----------------------

Hello, world!
```
I'll add a comment to say we are retrying the actual argument.

Reply:
That's true: it's a redundant option in that case.

Reply:
Not if it was `-XX:-HeapDumpOnOutOfMemoryError`: it will be ignored and dumps will be enabled (the default).

Reply:
Right. We never enter the loop.  Will fix...

Reply:
Fixed.  @keithc-ca would you kindly review?
Thanks